orttraining/orttraining/lazy_tensor/bridge.cc

```diff
@@ -0,0 +1,295 @@
+      ORT_ENFORCE(device.Id() >= 0, "ORT GPU device ID must be >= 0 but got ", device.Id());
+      // c10 GPU can have negative index (means current device),
+      // but only using non-negative index is enough to cover all ORT cases.
+      return c10::Device(c10::DeviceType::CUDA, device.Id());
+    }
+    default: {
+      // Got unsupported device. Throws.
+      const char* device_str = nullptr;
+      if (device.Type() == OrtDevice::CPU) {
+        device_str = "CPU";
+      } else if (device.Type() == OrtDevice::GPU) {
+        device_str = "GPU";
+      } else {
+        device_str = "Unknown";
+      }
+      ORT_THROW("Unsupport ORT device: ", device_str, ", ID: ", device.Id());
+    }
+  }
+}
+
+OrtValue CreateOrtTensorValue(const at::Tensor& tensor) {
+  onnxruntime::MLDataType element_type = CreateOrtScalarType(tensor.scalar_type());
+  onnxruntime::TensorShape shape(tensor.sizes().vec());
+  OrtDevice device = CreateOrtDevice(tensor.device());
+  OrtMemoryInfo memory_info = OrtMemoryInfo("LTC", OrtAllocatorType::OrtDeviceAllocator, device, device.Id());
+  // This tensor's life time is controlled by Pytorch.
+  // TODO: consider to let ORT also own that tensor.
+  std::unique_ptr<onnxruntime::Tensor> ort_tensor = std::make_unique<onnxruntime::Tensor>(
+      element_type, shape,
+      tensor.data_ptr(), memory_info);
+
+  OrtValue ort_value;
+  ort_value.Init(
+      ort_tensor.release(),
+      onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>(),
+      onnxruntime::DataTypeImpl::GetType<onnxruntime::Tensor>()->GetDeleteFunc());
+  return ort_value;
+}
+
+c10::IValue CreateC10IvalueTensor(OrtValue value) {
+  onnxruntime::Tensor* tensor = value.GetMutable<onnxruntime::Tensor>();
+  const OrtDevice& device = tensor->Location().device;
+  auto options = torch::TensorOptions()
+                     .dtype(CreateC10ScalarType(tensor->DataType()->AsPrimitiveDataType()))
+                     .layout(torch::kStrided)
+                     .device(CreateC10Device(device))
+                     .requires_grad(false);
+
+  // Extract shape from onnxruntime::TensorShape as a vector.
+  auto create_shape_vector = [](const onnxruntime::TensorShape& shape) {
+    std::vector<int64_t> new_shape(shape.NumDimensions());
+    shape.CopyDims(new_shape.data(), shape.NumDimensions());
+    return new_shape;
+  };
+
+  at::Tensor new_tensor = torch::from_blob(
+      tensor->MutableDataRaw(),
+      create_shape_vector(tensor->Shape()),
+      // Capture-by-value means
+      //  1. A new OrtValue is direct-initialized from "value".
+      //  2. The new OrtValue and "value" share the same underlying tensor, so
+      //     the tensor's lifetime is controlled by both of them, whichever is longer.
+      //  3. The new OrtValue's lifetime is the same as this lambda function.
+      //  4. This lambda function is deleted by "new_tensor"'s dtor, which also ends
+      //     the underlying tensor's life.
+      [value, tensor](void* p) {
```

Comment:
Basically, you passed a shared_ptr by value. https://docs.microsoft.com/en-us/cpp/cpp/how-to-create-and-use-shared-ptr-instances?view=msvc-170 has a very good explanation of this. I still don't understand what this code is for. Especially, it is an empty function.